File: 문제풀이/chap3-6/prob3-6-3.py
def MergeSort(A):
	# 길이가 1이라면 이미 정렬되어 있으므로, 아무것도 하지 않음
	if len(A) == 1:
		return A
	
	# 2개로 분할하고, 분할한 배열을 정렬함
	m = len(A) // 2
	A_Dash = MergeSort(A[0:m])
	B_Dash = MergeSort(A[m:len(A)])
	
	# 이 시점에서 다음 두 배열이 정렬 완료됨
	# 배열 A' [A_Dash[0], A_Dash[1], ..., A_Dash[m-1]]
	# 배열 B' [B_Dash[0], B_Dash[1], ..., B_Dash[len(A)-m-1]]
	# 다음 코드들은 Merge 조작에 해당하는 부분
	c1 = 0
	c2 = 0
	C = []
	while (c1 < len(A_Dash) or c2 < len(B_Dash)):
		if c1 == len(A_Dash):
			# 배열 A'가 빈 경우
			C.append(B_Dash[c2])
			c2 += 1
		elif c2 == len(B_Dash):
			# 배열 B'가 빈 경우
			C.append(A_Dash[c1])
			c1 += 1
		else:
			# 둘 다 아닌 경우
			if A_Dash[c1] <= B_Dash[c2]:
				C.append(A_Dash[c1])
				c1 += 1
			else:
				C.append(B_Dash[c2])
				c2 += 1
	
	# 배열 A', 배열 B'를 Merge한 배열 C를 리턴
	return C

# 메인 부분
N = int(input())
A = list(map(int, input().split()))

# 병합정렬 → 답 출력
Answer = MergeSort(A)
print(*Answer)

File: 문제풀이/chap3-6/prob3-6-3.c
#include <stdio.h>

int N, A[200009], C[200009];

// A[l], A[l+1], ..., A[r−1]를 오름차순으로 정렬하는 함수
void MergeSort(int l, int r) {
	// r−l=1이라면 이미 정렬되어 있으므로, 아무것도 하지 않음
	if (r - l == 1) return;
	
	// 2개로 분할하고, 분할한 배열을 정렬함
	int m = (l + r) / 2;
	MergeSort(l, m);
	MergeSort(m, r);
	
	// 이 시점에서 다음 두 배열이 정렬 완료됨
	// 배열 A' [A[l], A[l+1], ..., A[m−1]]
	// 배열 B' [A[m], A[m+1], ..., A[r−1]]
	// 다음 코드들은 Merge 조작에 해당하는 부분
	int c1 = l, c2 = m, cnt = 0;
	while (c1 != m || c2 != r) {
		if (c1 == m) {
			// 배열 A'가 빈 경우
			C[cnt] = A[c2]; c2++;
		}
		else if (c2 == r) {
			// 배열 B'가 빈 경우(빠진 부분)
			C[cnt] = A[c1]; c1++;
		}
		else {
			// 모두 아닌 경우(빠진 부분)
			if (A[c1] <= A[c2]) {
				C[cnt] = A[c1]; c1++;
			}
			else {
				C[cnt] = A[c2]; c2++;
			}
		}
		cnt++;
	}
	
	// 배열 A', 배열B'를 Merge한 배열 C를 A에 하나하나 할당
	// [C[0], ..., C[cnt−1]] −> [A[l], ..., A[r−1]]
	for (int i = 0; i < cnt; i++) A[l + i] = C[i];
}

int main() {
	// 입력
	scanf("%d", &N);
	for (int i = 1; i <= N; i++) scanf("%d", &A[i]);
	
	// 병합정렬 → 답 출력
	MergeSort(1, N + 1);
	for (int i = 1; i <= N; i++) printf("%d\n", A[i]);
	return 0;
}